src/python/pants/backend/terraform/tffmt.py

```diff
@@ -0,0 +1,198 @@
+
+    sources: TerraformSources
+
+
+class TffmtRequest(TerraformFmtRequest):
+    field_set_type = TffmtFieldSet
+
+
+@dataclass(frozen=True)
+class SetupRequest:
+    request: TffmtRequest
+    check_only: bool
+
+
+@dataclass(frozen=True)
+class Setup:
+    directory_to_process: Dict[str, Process]
+    original_digest: Digest
+
+
+@rule(level=LogLevel.DEBUG)
+async def setup_terraform_fmt(setup_request: SetupRequest, terraform: TerraformTool) -> Setup:
+    download_terraform_request = Get(
+        DownloadedExternalTool,
+        ExternalToolRequest,
+        terraform.get_request(Platform.current),
+    )
+
+    source_files_request = Get(
+        SourceFiles,
+        SourceFilesRequest(field_set.sources for field_set in setup_request.request.field_sets),
+    )
+
+    downloaded_terraform, source_files = await MultiGet(
+        download_terraform_request, source_files_request
+    )
+
+    source_files_snapshot = (
+        source_files.snapshot
+        if setup_request.request.prior_formatter_result is None
+        else setup_request.request.prior_formatter_result
+    )
+
+    input_digest = await Get(
+        Digest,
+        MergeDigests((source_files_snapshot.digest, downloaded_terraform.digest)),
+    )
+
+    # `terraform fmt` operates on a directory-by-directory basis. First determine the directories in
+    # the snapshot. This does not use `source_files_snapshot.dirs` because that will be empty if the files
+    # are in a single directory.
+    directories = defaultdict(list)
+    for file in source_files.snapshot.files:
+        directory = os.path.dirname(file)
+        if directory == "":
+            directory = "."
+        directories[directory].append(file)
+
+    # Then create a process for each directory.
+    directory_to_process = {}
+    for directory, files_in_directory in directories.items():
+        args = [
+            "./terraform",
+            "fmt",
+        ]
+        if setup_request.check_only:
+            args.append("-check")
+        args.append(directory)
+
+        process = Process(
+            argv=args,
+            input_digest=input_digest,
+            output_files=files_in_directory,
+            description=f"Run `terraform fmt` on {pluralize(len(files_in_directory), 'file')}.",
+            level=LogLevel.DEBUG,
+        )
+
+        directory_to_process[directory] = process
+
+    return Setup(
+        directory_to_process=directory_to_process, original_digest=source_files_snapshot.digest
+    )
+
+
+@rule(desc="Format with `terraform fmt`")
+async def tffmt_fmt(request: TffmtRequest, tffmt: TfFmtSubsystem) -> FmtResult:
+    if tffmt.options.skip:
+        return FmtResult.skip(formatter_name="tffmt")
+    setup = await Get(Setup, SetupRequest(request, check_only=False))
+    processes = [
+        (directory, Get(ProcessResult, Process, process))
+        for directory, process in setup.directory_to_process.items()
+    ]
+    results = await MultiGet(p[1] for p in processes)
```

Comment:
Dictionaries are ordered now in Python 3.6+, so there's no need for this intermediate type. You can do 

```python
results = await MultiGet(Get(ProcessResult, Process, process) for process in setup.directory_to_process.values())
```

Then down below:

```python
for directory, result in zip(setup.directory_to_process.keys(), results)
```